### PR TITLE
[AIRFLOW-1665] Database reconnect logic

### PR DESCRIPTION
**Pull request by vklogin**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-1665) issues and references them in the PR title. 

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
Currently when the scheduler gets a database error, the  scheduler errors out.  However, sometimes the database outage is a temporary network issue or a short lived outage that the server should try reconnect.  As part of this change, a new configuration entry is added allowing for specifying how to retry reconnecting to the database as well as us reconnect code for sqlalchemy.

### Tests
- [X] My PR does not need testing for this extremely good reason: The code added is for database reconnect/retry.  It cannot be tested using unit tests.

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/utils/db.py [hunk 1]
```diff
@@ -21,13 +21,17 @@
 from functools import wraps
 import logging
 import os
+import time
+
+import sys, traceback
```
Comment (bolkedebruin) [inline airflow/utils/db.py]: seems unused

**Comment by StephanErb** [comment]
I am slightly confused. Isn't this already mostly addressed (in the existing code) here?https://github.com/apache/incubator-airflow/blob/034b7ede9a25a9bddbe94e37f62aaaeb2fe0ada5/airflow/utils/db.py#L59-L71

**Comment by vklogin** [comment]
@StephanErb sqlalchemy documents say that when a DisconnectError is raised, it apparently retries.  But I did not observe this - so either it does not retry, or it gives up too quickly.  So pessimistic_connection_handling() event handler does not help.  Just checked the doc - DisconnectError() retries three times before giving up.  http://docs.sqlalchemy.org/en/latest/core/exceptions.html#sqlalchemy.exc.DisconnectionError

In our case, the My SQL goes down for a few seconds during occasional maintenance and this was causing all the schedulers to exit.

**Comment by vklogin** [comment]
Can someone please re-trigger the travis build?  I see it's failing with:
`ERROR: InvocationError: '/home/travis/build/apache/incubator-airflow/scripts/ci/check-license.sh'
___________________________________ summary ____________________________________
ERROR:   py34-cdh-airflow_backend_postgres: commands failed`
Same issue was also reported in https://github.com/apache/incubator-airflow/pull/2431#issuecomment-314705911

**Comment by bolkedebruin** [comment]
You need to do an empty commit (rebase). We can’t trigger the build unfortunately.

File: airflow/utils/db.py [hunk 2]
```diff
@@ -59,6 +61,113 @@ def wrapper(*args, **kwargs):
...
+                    raise
+            else:
+                logging.warning("reconnect_on_engine_connect(): Not retrying as connection_invalidated is False")
+                raise
+        finally:
+            # restore "close with result"
+            logging.debug("reconnect_on_engine_connect() done")
+            connection.should_close_with_result = save_should_close_with_result
+
+    @event.listens_for(Pool, "checkout")
```
Comment (StephanErb) [inline airflow/utils/db.py:106]: If I understand http://docs.sqlalchemy.org/en/rel_1_1/core/pooling.html#disconnect-handling-pessimistic correctly then this is only needed when a pool without an engine is used. I believe this is not the case so the `engine_connect` handler above should be sufficient.

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -858,6 +858,19 @@ def scheduler(args):
         num_runs=args.num_runs,
         do_pickle=args.do_pickle)
 
+    db_reconnect_limit = 0
+    try:
+        db_reconnect_limit = conf.getint('scheduler', 'database_reconnect_limit')
+    except airflow.exceptions.AirflowConfigException as exc:
+        print("Unable to read database_reconnect_limit config; assuming no reconnect")
```
Comment (StephanErb) [inline airflow/bin/cli.py:865]: I am new to Airflow, so please excuse my ignorance: When can it happen that the value is not defined when it is part of the default config?

File: airflow/bin/cli.py [hunk 4]
```diff
@@ -858,6 +858,19 @@ def scheduler(args):
...
+
+    if (db_reconnect_limit == 0):
+        print("db_reconnect_limit is 0; disabling scheduler db reconnect")
+    else:
+        adjusted_db_reconnect_limit = db_reconnect_limit if db_reconnect_limit > 0 else None
+        print("db_reconnect_limit: ", adjusted_db_reconnect_limit)
+        db_utils.enable_connection_reconnect(adjusted_db_reconnect_limit)
```
Comment (StephanErb) [inline airflow/bin/cli.py:872]: `settings.engine` can change and is re-initialized for subprocesses. I therefore think the correct place for the handler setup is within `settings.configure_orm()`.

File: airflow/utils/db.py [hunk 5]
```diff
@@ -59,6 +61,113 @@ def wrapper(*args, **kwargs):
...
+
+    if succeeded:
+        logging.info("retry_connection_for_period(): Connection re-established")
+        return True
+
+    logging.error("retry_connection_for_period(): Connection could NOT be re-established")
+    return False
+
+
 def pessimistic_connection_handling():
```
Comment (StephanErb) [inline airflow/utils/db.py:171]: With the new code above, this should be save to remove.

**Comment by StephanErb** [comment]
Short heads-up: I am currently struggling with a flaky DB at work that takes down Airflow instances way to easily. I am working on a set of patches that also includes a variant of this patch here as its groundwork.

I will test this tomorrow and report back with my findings.

In case you are interested in the current work in progress:
* fixed pool ping: https://github.com/apache/incubator-airflow/commit/cda2bed04c23437890ec6388b8c58ddd85607afe
* reconnect logic: https://github.com/apache/incubator-airflow/commit/6884f38ea48b99c48fda28b6c7edf5bf251c68f9
* full diff (best read commit by commit)  https://github.com/apache/incubator-airflow/compare/master...StephanErb:db_robustness

File: airflow/bin/cli.py (hunk 3 above)
Comment (Fokko) [reply airflow/bin/cli.py:865]: It will throw an exception that the configuration key cannot be found

**Comment by StephanErb** [comment]
@vklogin I have just pushed a similar patch https://github.com/apache/incubator-airflow/pull/2822. Would be great if you could take a look if this would also work for you. Thanks!

**Comment by bolkedebruin** [comment]
Is this still relevant? Can you make sure to base against master in case it is or otherwise close it?

**Comment by vklogin** [comment]
https://github.com/apache/incubator-airflow/pull/2822 appears to include this change.  Will verify it handles our issues and Close this PR.

**Comment by bolkedebruin** [comment]
@vklogin any update?

**Comment by r39132** [comment]
Since we haven't heard back, I'm closing this.

**Comment by henrym2** [comment]
We're still having this issue on Airflow 1.10.3, the issue is seemingly causing Airflow to eventually crash as it loses connection to the main "Airflow" database.

**Comment by ryw** [comment]
@henrym2 which component(s) of Airflow are you experiencing this issue with?

**Pull request closed without merging**